### PR TITLE
MultiplePriorityQueues: Add ability to POP first message from a list of queues

### PR DESCRIPTION
**Pull request by ofirnk**
This PR adds ability to provide a list of queues 
`[first_priority, second_priority, third_priority, ...]`
 and receive the first message available.

This is done via a single _Redis_ command that iterates **_server-side**_ on the provided list of queues.

Suggestions how to improve this PR are welcome :tada:

### CODE REVIEW

File: src/main/java/net/greghaines/jesque/worker/WorkerImpl.java [hunk 1]
```diff
@@ -36,24 +49,32 @@
...
+import static net.greghaines.jesque.worker.WorkerEvent.JOB_EXECUTE;
+import static net.greghaines.jesque.worker.WorkerEvent.JOB_FAILURE;
+import static net.greghaines.jesque.worker.WorkerEvent.JOB_PROCESS;
+import static net.greghaines.jesque.worker.WorkerEvent.JOB_SUCCESS;
+import static net.greghaines.jesque.worker.WorkerEvent.WORKER_ERROR;
+import static net.greghaines.jesque.worker.WorkerEvent.WORKER_POLL;
+import static net.greghaines.jesque.worker.WorkerEvent.WORKER_START;
+import static net.greghaines.jesque.worker.WorkerEvent.WORKER_STOP;
+import static net.greghaines.jesque.worker.WorkerImpl.NextQueueStrategy.DRAIN_WHILE_MESSAGES_EXISTS;
+import static net.greghaines.jesque.worker.WorkerImpl.NextQueueStrategy.RESET_TO_HIGHEST_PRIORITY;
```
Comment (gresrun) [inline src/main/java/net/greghaines/jesque/worker/WorkerImpl.java]: Please keep the import order that the rest of the project uses:
<static imports> (If more than a few from the same class, import star)
<regular imports>

Comment (ofirnk) [inline src/main/java/net/greghaines/jesque/worker/WorkerImpl.java]: right, sorry about that :+1:

**Comment by ofirnk** [comment]
On latest changeset I've added `setOrderedPriorityQueues(List queues)` to `Worker` interface

**Comment by shlomisut** [comment]
@gresrun Can you have a look here ?

**Comment by rathboma** [comment]
Hey, we're just starting to use Jesque to replace our horrible mess of futures and threads, and saw this pull request. Just curious if this functionality is going to make it in? Looks like exactly what we need.

**Comment by gresrun** [comment]
I'll try to take a look at this tonight.

**Comment by shlomisut** [comment]
@rathboma from our end we can say we've been using this in production for the last few months, and it has been working great.

@gresrun thanks, will be happy to see this as part of the main artifact!

File: src/main/java/net/greghaines/jesque/worker/WorkerImpl.java [hunk 2]
```diff
@@ -185,9 +208,9 @@ public void run() {
                 this.jedis.sadd(key(WORKERS), this.name);
                 this.jedis.set(key(WORKER, this.name, STARTED), new SimpleDateFormat(DATE_FORMAT).format(new Date()));
                 this.listenerDelegate.fireEvent(WORKER_START, this, null, null, null, null, null);
-                this.popScriptHash.set(this.jedis.scriptLoad(ScriptUtils.readScript("/workerScripts/jesque_pop.lua")));
-                this.lpoplpushScriptHash.set(this.jedis.scriptLoad(
-                        ScriptUtils.readScript("/workerScripts/jesque_lpoplpush.lua")));
+                this.popScriptHash.set(this.jedis.scriptLoad(ScriptUtils.readScript(POP_LUA)));
+                this.lpoplpushScriptHash.set(this.jedis.scriptLoad(ScriptUtils.readScript(LPOPLPUSH_LUA)));
+                this.multiPriorityQueuesScriptHash.set(this.jedis.scriptLoad(ScriptUtils.readScript(LPOPLPUSH_LUA)));
```
Comment (gresrun) [inline src/main/java/net/greghaines/jesque/worker/WorkerImpl.java]: I'm not following this logic. Why have two script hashes for loading the same script, LPOPLPUSH_LUA? Is this second one supposed to load workerScripts/fromMultiplePriorityQueues.lua?

File: src/main/java/net/greghaines/jesque/worker/WorkerImpl.java [hunk 3]
```diff
@@ -328,6 +351,14 @@ public void removeAllQueues() {
...
     @Override
     public void setQueues(final Collection<String> queues) {
+        setOrderedPriorityQueues(new ArrayList<>(queues));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setOrderedPriorityQueues(final List<String> queues) {
```
Comment (gresrun) [inline src/main/java/net/greghaines/jesque/worker/WorkerImpl.java]: Not sure I see the rationale behind making this a List argument. You aren't really using it as a list, just adding all the values into the BlockingDeque which all you need is a Collection for.

File: src/test/java/net/greghaines/jesque/worker/TestWorkerPool.java [hunk 4]
```diff
@@ -17,6 +9,15 @@
...
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
```
Comment (gresrun) [inline src/test/java/net/greghaines/jesque/worker/TestWorkerPool.java]: Same here about the imports.

File: src/test/java/net/greghaines/jesque/worker/TestWorkerImpl.java [hunk 5]
```diff
@@ -1,21 +1,27 @@
...
 import org.junit.Assert;
 import org.junit.Test;
+import redis.clients.jedis.Jedis;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+
+import static net.greghaines.jesque.utils.JesqueUtils.entry;
+import static net.greghaines.jesque.utils.JesqueUtils.map;
```
Comment (gresrun) [inline src/test/java/net/greghaines/jesque/worker/TestWorkerImpl.java]: Please keep the import order consistent with the rest of the project.

File: src/main/java/net/greghaines/jesque/worker/WorkerImpl.java [hunk 6]
```diff
@@ -440,15 +473,43 @@ protected void poll() {
...
     protected String pop(final String curQueue) {
         final String key = key(QUEUE, curQueue);
-        return (String) this.jedis.evalsha(this.popScriptHash.get(), 3, key, key(INFLIGHT, this.name, curQueue), 
-                JesqueUtils.createRecurringHashKey(key), Long.toString(System.currentTimeMillis()));
+        switch (nextQueueStrategy) {
+            case DRAIN_WHILE_MESSAGES_EXISTS:
+                return (String) this.jedis.evalsha(this.popScriptHash.get(), 3, key, key(INFLIGHT, this.name, curQueue),
+                        JesqueUtils.createRecurringHashKey(key), Long.toString(System.currentTimeMillis()));
+            case RESET_TO_HIGHEST_PRIORITY:
+                return (String) this.jedis.evalsha(this.multiPriorityQueuesScriptHash.get(), 1, curQueue);
```
Comment (gresrun) [inline src/main/java/net/greghaines/jesque/worker/WorkerImpl.java]: This shouldn't work as written. See my first comment.

File: src/test/java/net/greghaines/jesque/worker/TestWorkerImpl.java (hunk 5 above)
Comment (ofirnk) [inline src/test/java/net/greghaines/jesque/worker/TestWorkerImpl.java]: right - will change

File: src/test/java/net/greghaines/jesque/worker/TestWorkerPool.java (hunk 4 above)
Comment (ofirnk) [inline src/test/java/net/greghaines/jesque/worker/TestWorkerPool.java]: will change also

**Comment by bp-FLN** [comment]
Any reason why this feature has not been added to the WorkerPoolImpl?
Are there any technical issues that prevent this from being done?
I'd like to provide a PR but need to make sure first if it's worth implementing.

**Comment by ofirnk** [comment]
AFAICR we just didn't need it there, and this implementation was enough for our use case

**Pull request merged**